### DISCUSSION TITLE
Manipulating CAN data on charging current and SOC for better charging scheme of the Inverter

### DISCUSSION THREAD

**Discussion by paulsteigel** [ArminJo/JK-BMSToPylontechCAN#19] [Ideas]
Hi Armin,
With some studying and observation over charging scheme of the Inverter (Luxpower - mine) I discovered that we can manipulate CAN response to the Inverter so it will reduce charging rush when the battery reaches 100%. See image below:
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/46605996-70b1-40b2-8765-a6c3f504bd79)
So we will do check the SOC during the charging:
1. When it reaches 80% SOC >> start reducing charging current (the easiest shall be gradual reduction of current to match with SOC's increasing);
2. When SOC is 100% reduced the continuous charging current to approx. 0A
This approach worked with ESPhome as I have sent you over email sometime ago.
If you would have sometimes, please extend your help for improving the code with thanks.
Best!
Ngoc

**Comment by ArminJo** [comment]
What is the Y axis? Current?
And adaptive charging will prolonge the time to reach 100%.

**Reply by paulsteigel** [reply]
Hi Armin,
Please see the full graph here attached. Y axis if for Wattage
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/7593074c-f0f2-43fd-b9e3-caa96121f4ce)
The newer SNA5000 does have this reduced charging scheme but rush increasing of charging power took place sometime where the SOC is already 100% >> second image
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/c4dbae53-fd0a-4712-a666-68dc88fde9bf)
But this one with Standard protocol is nicely done. So i doubt CAN message was manipulated by the battery BMS (see image 3)
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/3023ec24-7897-46c7-b991-2c3b7aaa3af2)
All these graphs were taken from live view of my families inverter over web. the first one was taken from older inverter (2022 version) and the rest are newly purchased in 2023 (July to now)

Uksa007 applied a different approach with timing, but I think, we do not need timing, just use SOC and/or Voltage to determine the appropriate charging current. Such as a equation to reduce current based on the increasing SOC/Voltage.
I think the option on charging voltage that Ukas007 applied can be of help, the equation is somewhat like:
Can_charging_voltage = (Actual_charging_voltage - 2.5) * 10 
Ref: https://github.com/Uksa007/esphome-jk-bms-can/blob/main/esp32-example-can.yaml
 
Best
Ngoc

**Reply by ArminJo** [reply]
Hi, the last two links are broken.
Di you have Li-Ion or LiFePO batttery?
I implemented a user function. https://github.com/ArminJo/JK-BMSToPylontechCAN#compile-options--macros-for-this-software -> scroll down. 
Can you please test it.

**Reply by paulsteigel** [reply]
Hi Armin,
Many thanks for your great works!
My original writings was a bit irresponsible (just only citing out without any consideration for proposing at all), I rewrote for easy going. Why waiting for your solution, I have tested from my side with modification on the Pylontech_CAN.h at PylontechCANBatteryLimitsFrameStruct section as below:
====START CODE==========
struct PylontechCANBatteryLimitsFrameStruct {
    struct PylontechCANFrameInfoStruct PylontechCANFrameInfo = { PYLON_CAN_BATTERY_LIMITS_FRAME_ID, 8 }; // 0x351
    struct {
        int16_t BatteryChargeOvervoltage100Millivolt;       // 0 to 750
        int16_t BatteryChargeCurrentLimit100Milliampere;    // 0 to 5000
        int16_t BatteryDischargeCurrentLimit100Milliampere; // -5000 to 0
        int16_t BatteryDischarge100Millivolt;               // 0 to 65535 // not in documentation
    } FrameData;
    void fillFrame(struct JKReplyStruct *aJKFAllReply) {
      // Checking Soc and then manipulate current/ voltage for stop charging or apply a smooth charging scheme here
      // Needed for old Luxpower and also avoid additional charging
      if (aJKFAllReply->SOCPercent == 100 || swap(aJKFAllReply->ChargeOvercurrentProtectionAmpere) * 10 > 0){
        //Stop charging and discharging for now
        FrameData.BatteryChargeCurrentLimit100Milliampere =0;
        FrameData.BatteryDischargeCurrentLimit100Milliampere = 0;
      } else if (aJKFAllReply->SOCPercent <= 85){
        // Full charging mode as set by BMS
        //JKComputedData.Battery10MilliAmpere = getCurrent(sJKFAllReplyPointer->Battery10MilliAmpere);
        //JKComputedData.BatteryLoadCurrentFloat = JKComputedData.Battery10MilliAmpere;
        //JKComputedData.BatteryLoadCurrentFloat /= 100;

        FrameData.BatteryChargeOvervoltage100Millivolt = JKComputedData.BatteryFullVoltage10Millivolt / 10;
        FrameData.BatteryChargeCurrentLimit100Milliampere = swap(aJKFAllReply->ChargeOvercurrentProtectionAmpere) * 10;
        FrameData.BatteryDischargeCurrentLimit100Milliampere = swap(aJKFAllReply->DischargeOvercurrentProtectionAmpere) * 10;
        FrameData.BatteryDischarge100Millivolt = swap(aJKFAllReply->BatteryUndervoltageProtection10Millivolt) / 10;
      } else if (aJKFAllReply->SOCPercent > 85) {
        FrameData.BatteryChargeCurrentLimit100Milliampere = (1/aJKFAllReply->SOCPercent)* swap(aJKFAllReply->ChargeOvercurrentProtectionAmpere) * 10;
        FrameData.BatteryDischargeCurrentLimit100Milliampere = 0;
      } 
    }
};
==END CODE=========


And with the above modification, the inverter stops additional charging at 85%. Therefore, I believe the approach would work. I will test your Macro tonight and response.
Initially, with my above quick test, the Inverter will display BMS Max charging current limit: 0 A which is a bit weird. After a bit of research applying control on voltage shall be better (see UKsa007 (https://github.com/Uksa007/esphome-jk-bms-can/blob/main/esp32-example-can.yaml) - line 213 on ward. I will try your approach first and will áo modify a bit my previouls test code to change from current limit to charging voltage.
The algorithm will somewhat like this:
1. Check if battery is charging, check SOC
2. If charging >>  check if SOC is 80% or above
3. If yes, start to reduce voltage incrementally from current charging voltage to floating voltage (final floating = cell OPV * number of cells)
UKsa007 using charging voltage from 52.8 to 57.6 for 16S LFP battery  3.3-3.6V for cell, incremental step of 0.1 V. So we can take the OVPR as the maximum and 3.3 * number of cell as minimum. Uksa007 using timing for charging voltage slope, we should use SOC as reference to define the slope would be useful. For each % increasing of SOC in the range of 80-100%, charging voltage will be reduced accordingly regardless how long charging will take place. 
The reducing charging power is especially needed for lithium battery (4.2V max).
Please advice! 
Again, thanks for your efforts.
Best
Ngoc

**Reply by paulsteigel** [reply]
Hi Armin,
I tested the new setting and the charge reduced successfully as required with a bit of problem on reaching 100% SOC. The current reduced gradually and cause difficulties for the the battery to reach 100%,
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/4cf13950-1cff-4ed5-b753-508dbb7f74e0)

![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/8855c361-f93a-452d-9606-2c2cf4790130)

There is one problem that the BMS will report Max charging current changed over time (that I dont see from the standard BMS - image below)
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/d243971f-73dd-4a12-8ec0-f04bbd173ca3).
So, as I mentioned in the last comment, charging voltage shall be of suitable solution (so the Inverter will not see the changes in charging curent) with a few changes as following:
Input parameter
+ Start linear Charge SOC: 80%
+ End of linear charge SOC: 96%: This is to allow a strong enough current to make the battery to reach 100% or at least 99%
1. Check if we are in charging mode
2. Yes, Check SOC 
3. Case 1: Below 80%: no modification
4. Case 2: between 80-96%:  Apply linear reduction of charging voltage from the current charging voltage to floating charging voltage:
+ Voltage at 80%: OverVoltageProtection * Number_of_Cell (V1)
+ Voltage at 96%: ~ float voltage (V2)
Step of reduction is based on the map(V1,V2) as you have specified in your code.
I think this would be nice to go.
Many thanks for your great works.
If possible, the above option will also be good for increasing charging current that is good for battery and also at the time that Solar power is still low (staging as in the image below).
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/2dae1257-f06d-4bf4-aeb9-7e8add18bedf)
Start to ramp up at 60% after two hours from 8AM.
![image](https://github.com/ArminJo/JK-BMSToPylontechCAN/assets/8839597/191370fa-6bca-4e30-9d73-8ff839f09a1a)
So the map for charging voltage increase shall be:
Map(Floating Charge, OvervoltageProtection) for the SOC range from 60-80
Best! 
Ngoc
Best Ngoc

**Reply by ArminJo** [reply]
> between 80-96%: Apply linear reduction of charging voltage from the current charging voltage to floating charging voltage:
Why dont you change the value of MAX_CURRENT_MODIFICATION_MIN_CURRENT_TENTHS_OF_AMPERE?
It is almost equivalent to your proposition.
What do you mean with `Map(Floating Charge, OvervoltageProtection) for the SOC range from 60-80`? This is no map function!

**Reply by paulsteigel** [reply]
Hi Armin,
Thanks for your response, I understand your point, current limiting also works but Inverter will report it on the web interface, so controll charging voltage will be better as the Inverter will not report it on the screen!

> Why dont you change the value of MAX_CURRENT_MODIFICATION_MIN_CURRENT_TENTHS_OF_AMPERE?

> What do you mean with Map(Floating Charge, OvervoltageProtection) for the SOC range from 60-80? This is no map function!

This is based on your code
`PylontechCANBatteryLimitsFrame.FrameData.BatteryChargeCurrentLimit100Milliampere = map(sJKFAllReplyPointer->SOCPercent,
                MAX_CURRENT_MODIFICATION_LOWER_SOC_THRESHOLD_PERCENT, 100,
                PylontechCANBatteryLimitsFrame.FrameData.BatteryChargeCurrentLimit100Milliampere,
                MAX_CURRENT_MODIFICATION_MIN_CURRENT_TENTHS_OF_AMPERE);`

What I tried to reiterate is to make it simple (my English is a bit bad).
Please understand, i am trying to learn your code as well. Some parts were out of my existing knowledge. Therefore, It better to illustrate the context better for you to help.
Many thanks for your great efforts!
Best, 
Ngoc
